### ISSUE TITLE
Feature request : Improve automatic tag detection

### ISSUE THREAD

**Issue by 3like3beer** [jflesch/paperwork#445]
Hi,

I have been using paperwork for a few weeks and I am quite happy with it. So thanks.
I really use it as a scan and forget routine with automatic encrypted back-ups in case my PC crashes.

I tried the tag feature but the tags guess were ... not that good.
So I removed former tags and I text-search when I have something to find. It's the kind of strategy I have with my e-mails (archive everything not sorted and text search if I need to find something back, it works for me).

However I think tagging can be great, especially if I don't have to do it at all.
Have you tried something like document clustering (http://brandonrose.org/clustering http://scikit-learn.org/stable/auto_examples/text/document_clustering.html) ?
If not do you think it can be envisioned ?
I have some statistical background (more with R and SAS) and a bit of python knowledge, but I don't have a lot of time. However if you think it is conceivable to have such a feature I can try to prototype it.
Brgds,

Julien

**Comment by jflesch** [comment]
1) With how many documents did you try ? I have 1200, and the results of the label guessing are pretty good. However, I haven't tested it much with a small set of document.

2) On the principle, I have nothing against using a method other than Bayesian filters.
Just regarding scikit / numpy, they are really problematic to have as a dependency in Paperwork. Previously, label guessing was done using scikit / numpy (based on the text, but also on the thumbnail of the documents) (coded by someone else). It created many problems : It made installation much harder, it made packaging in distribution much harder, and it was quite slow.

3) I just had a very quick look at what you suggest. I have a (maybe stupid) question : It seems to many that document clustering is designed to sort documents into exclusive categories. Labels in Paperwork are not exclusive. I guess a solution would be to do as many document clusterings as there are labels (with each time only 2 categories: "has the label" / "has not the label"). But would that be efficient ? (more than the Bayesian filters I mean)

**Comment by 3like3beer** [comment]
Well ... fast reply.

1) Around 100, surely not enough

2) Ok, I think you're right. 
My title and my text was not clear. My idea is to have the tag defined automatically.
Use case would be scan, scan,... not bother with labelling, sorting... After some time run the algo to define/apply tags from OCRed text.
3) Clustering can be a way to do it and you are right the output of clustering is a partition. So it is not directly suitable for tags (even if you could "cut" the partitions at different levels and use the partition most relevent terms as labels, but this would be ... sloppy). However there are other methods (acp...) 
My point was not on the method/algo used (even if the issue title is misleading). 
The feature request would be better expressed as having tags defined (new feature) and applied (existing feature) automatically (scan and forget nothing else ... then retrieve if needed).
When I defined my tags I put : my children name / the companies issuing the document / the type of doc (invoice / contract)...
This is the kind of tags I think an algo could discover automatically.

Anyway, I currently don't have that many docs, so I am not sure that the idea is worth anything.

**Comment by jflesch** [comment]
1) I think it should be enough. I'll give a try when I have enough free time (probably not for a while ...)

2) Hm, I see the idea, but from a GUI standpoint, I don't think it would be easy to represent cleanly.

3) Ok, let's see if I understand what you're asking:
- You want to be able to define a list of labels, without attaching them immediatly to documents ?
- You want a button to guess the labels on all the already-scanned documents ?

>  the companies issuing the document

Just a quick suggestion : don't do that. With time, there will be a lot of companies, and you will end up with a lot of labels to manage. Since the OCR can usually extract the name of company from the document, it's pointless. If required, you can find them back quickly just by searching the name of the company. Worst case scenario, you can usually find them back by filtering by label and looking at the dates of the documents.
I tend to use only vague labels ("bills", "car", "house", etc), and let the OCR+search engine take care of the rest.
Remember, the goal of Paperwork is being lazy ;-)

**Comment by 3like3beer** [comment]
2 - It could a menu item be like the avancé - réindexer tous les documents
avancé - taggé les documents
3 - Not really :
I want to be able *not* to define labels. I want (well I'd like) to have a button (or another trigger) so that paperwork defines them by itself based on the text the existing documents have inside / and apply them to these documents.
4 - companies... Ok I guess you're right. Actually since I stopped using labels and remove the ones I had, it's fine now.

**Comment by Yusei** [comment]
Label guessing doesn't really work for me either. It worked for the first few scans, and was able to separate invoices from payslips. Now that I have around 100 documents scanned, it is no longer the case, and everything is always tagged with every label at first.

**Comment by jflesch** [comment]
Hm, interresting. Here something you can try: ``` rm -rf ~/.local/share/paperwork```
It will delete the index and bayesian cache used by Paperwork, forcing it to rebuild them from scratch next time it is started.

If it works, then it means I didn't manage the bayesian filters correctly.

Also, did you try with WEIGHT_YES = 1 ? I'm wondering if weighting the results was such a good idea.

**Comment by Yusei** [comment]
I scanned a document almost similar to one already in the database, tagged "work". With WEIGHT_YES=1, it got tagged with everything. I removed the document, removed the ~/.local/share/paperwork directory, and tried again. This time it correctly got tagged only with "work".

Maybe the bayesian predictors should only be trained when the user does something, instead of also being trained by their own results? I think that's how it works for spam detectors, but it would increase the false negatives a lot, I suppose.

**Comment by jflesch** [comment]
Hm, you're probably right. I will check, but I think right now, they train themselves (--> assume that the user would have corrected them if they did something wrong).
Was your last test still with WEIGHT=1 ?

**Comment by Yusei** [comment]
The last test was with WEIGHT_YES=5, WEIGHT_NO=1 (the default values).
Right now, they train themselves, and then they get trained again if the user changes something.

However, it seems that the correction only compensates for the incorrect training (one "no" cancels one incorrect "yes", but does not reinforce "no"). Now that I've retrained the bayesian predictors, it works for old tags, but whenever I add a new tag, it quickly goes to "always yes", until I retrain again.

I don't know if my suggestion of only training when the user does something is correct. Maybe a better way do to it would be to decide that a user action counts more than a default training (ie. you train twice if it is a user action). Anyway, it seems weird to train the predictor using its own prediction. If I have time I will simulate both strategies to see which one works best.

**Comment by jflesch** [comment]
Ok, I need to improve my test methodology.

Currently, I used ```scripts/stats.py``` what the best way to do things : It goes on each document, and see if the label guessed by Paperwork match the one actually on the document (using the current training).

I think I will have to write a script that simulate adding documents, starting from an empty work directory, + a user fixing the labels, and see what gives the best result.

**Comment by jflesch** [comment]
I've made a script : ```scripts/simulate-workdir.py```
It's now quite clear I'm not using Simplebayes correctly.

**Comment by Yusei** [comment]
Just a quick thought: regardless of the proper training strategy, maybe it would be possible to do a background check, from time to time, to see whether the predictors are correctly assigning the labels. If the error rate falls below a threshold, retrain them (assuming Simplebayes is fast enough so that it wouldn't be a problem for the user). That's what I'm doing by hand right now, it may not be very elegant but it works.

**Comment by jflesch** [comment]
Hm, the problems with that are:
- it is a workaround, and workaround are bad. Real solutions are much better. 
- and it would require a lot of code. Code which could work well .. statistically. And extra code means extra bugs

Anyway, when I destroy my index + bayesian filter caches and rebuild them from scratch, I get >95% success rate.
I should be able to get the same result by just adding documents like the user do. I think it's only a matter of fine-tuning the way I use of Simplebayes.

**Comment by jflesch** [comment]
Currently, the stats are really bad:
```
Nb documents:             517
Labels correctly guessed: 542/630 = 86%
Labels not guessed:       88/630 = 13%
Labels wrongly guessed:   1204/2200 = 54%
```
It gets wrong labels on pretty much all the documents.

**Comment by jflesch** [comment]
WEIGHT_YES=1.0, WEIGHT_NO=1.0, learn only from mistakes, no untraining before retraining
```
Nb documents:             549
Labels correctly guessed: 305/661 = 46%
Labels not guessed:       356/661 = 53%
Labels wrongly guessed:   81/397 = 20%
```

**Comment by jflesch** [comment]
@Yusel
Actually, there were bugs in my implementation :/ (Basically, when you changed a label in a document, the untraining / training wasn't really taken into account)
So maybe training the filters with the correctly-guessed labels is not to the problem here actually.

I'll keep doing tests.

**Comment by jflesch** [comment]
WEIGHT_YES = 1.0, WEIGHT_NO = 1.0, learn from all labels, learn only from first pages
```
Success/total:            438/931 = 47%
Labels correctly guessed: 531/1154 = 46%
Labels not guessed:       623/1154 = 53%
Labels wrongly guessed:   79/610 = 12%
```

**Comment by jflesch** [comment]
WEIGHT_YES = 1.0, WEIGHT_NO = 1.0, learn from all labels (guessed or from user), learn from whole document
```
Success/total:            326/736 = 44%
Labels correctly guessed: 450/901 = 49%
Labels not guessed:       451/901 = 50%
Labels wrongly guessed:   162/612 = 26%
```
(the success rate kept going down)

**Comment by jflesch** [comment]
WEIGHT_YES = 5.0, WEIGHT_NO = 1.0, learn from all labels (guessed or from user), learn from whole documents

```
Success/total:            490/826 = 59%
Labels correctly guessed: 800/1019 = 78%
Labels not guessed:       219/1019 = 21%
Labels wrongly guessed:   325/1125 = 28%
```

**Comment by jflesch** [comment]
WEIGHT_YES = 5.0, WEIGHT_NO = 1.0, learn from user only, learn from first page only

```
Success/total:            217/469 = 46%
Labels correctly guessed: 271/563 = 48%
Labels not guessed:       292/563 = 51%
Labels wrongly guessed:   72/343 = 20%
```

It started very well with a success rate > 60%, then it slowly kept falling down ...

**Comment by jflesch** [comment]
WEIGHT_YES = 2.5, WEIGHT_NO = 1.0, learn from user only, learn from first page only
```
Success/total:            523/951 = 54%
Labels correctly guessed: 738/1183 = 62%
Labels not guessed:       445/1183 = 37%
Labels wrongly guessed:   192/930 = 20%
```

It started also very well, but after ~200 documents, it settled at a success rate <= 55%.

**Comment by jflesch** [comment]
WEIGHT_YES = 2.5, WEIGHT_NO = 1.0, learn from all labels (guessed or from user) but with an emphasis on user changes (see branch wip-optim-learning-2), learn from first page only
```
Success/total:            324/608 = 53%
Labels correctly guessed: 446/730 = 61%
Labels not guessed:       284/730 = 38%
Labels wrongly guessed:   146/592 = 24%
```

**Comment by jflesch** [comment]
For reference, before the fixes, the result was the following:
```
Success/total:            4/102 = 3%
Labels correctly guessed: 103/117 = 88%
Labels not guessed:       14/117 = 11%
Labels wrongly guessed:   201/304 = 66%
```

I think I will have to do a release 0.3.1 asap.

**Comment by jflesch** [comment]
I think the version currently in the branch 'stable' is good enough. I'll release it as 0.3.1.
I'll see later if I can still improve it.
